File: nodejs/backend/src/service/LeadRequestService.ts
import { v4 as uuidv4 } from 'uuid';
import { LeadRequestRepository } from '../repository/LeadRequestRepository';
import { LeadRequest } from '../model/LeadRequest';
import { CreateLeadRequestDto } from '../dto/leadRequestDto';
import { LeadRequestNotFoundError } from '../error/AppError';

export class LeadRequestService {
  constructor(private readonly repo: LeadRequestRepository) {}

  async create(payload: CreateLeadRequestDto): Promise<LeadRequest> {
    const request: LeadRequest = {
      id: uuidv4(),
      name: payload.name,
      email: payload.email,
      phone: payload.phone,
      service: payload.service,
      message: payload.message,
      createdAt: new Date(),
    };

    await this.repo.insert(request);
    return request;
  }

  async list(): Promise<LeadRequest[]> {
    return this.repo.findAll();
  }

  async delete(id: string): Promise<void> {
    const deleted = await this.repo.deleteById(id);
    if (!deleted) {
      throw new LeadRequestNotFoundError();
    }
  }
}

File: rust/backend/src/service/lead_request_service.rs
use crate::{
    dto::CreateLeadRequest, error::AppError, model::LeadRequest,
    repository::LeadRequestRepository,
};
use chrono::Utc;
use std::sync::Arc;
use uuid::Uuid;
use validator::Validate;

#[derive(Clone)]
pub struct LeadRequestService {
    repo: Arc<dyn LeadRequestRepository>,
}

impl LeadRequestService {
    pub fn new(repo: Arc<dyn LeadRequestRepository>) -> Self {
        Self { repo }
    }

    pub async fn create(&self, payload: CreateLeadRequest) -> Result<LeadRequest, AppError> {
        payload.validate()?;

        let request = LeadRequest {
            id: Uuid::new_v4(),
            name: payload.name,
            email: payload.email,
            phone: payload.phone,
            service: payload.service,
            message: payload.message,
            created_at: Utc::now(),
        };

        self.repo.insert(&request).await?;
        Ok(request)
    }

    pub async fn list(&self) -> Result<Vec<LeadRequest>, AppError> {
        Ok(self.repo.find_all().await?)
    }

    pub async fn delete(&self, id: Uuid) -> Result<(), AppError> {
        match self.repo.delete(id).await? {
            Some(_) => Ok(()),
            None => Err(AppError::LeadRequestNotFound),
        }
    }
}
